mettle/api/api.go

```diff
@@ -554,55 +553,32 @@ func (s *server) process(msg *pubsub.Message) {
 				timeToComplete.Observe(j.LastUpdate.Sub(j.StartTime).Seconds())
 			}
 			log.Info("Job %s completed by %s", key, worker)
-			go s.deleteJob(key, j)
 		}
 	}
 }
 
-// deleteJob waits for a period then removes the given job from memory.
-func (s *server) deleteJob(hash string, j *job) {
-	time.Sleep(retentionTime + time.Duration(rand.Int63n(int64(retentionTime))))
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-	// Check the action hasn't been replaced since deleteJob was called
-	if s.jobs[hash] == j {
-		log.Notice("Removing job %s", hash)
-		delete(s.jobs, hash)
-		currentRequests.Dec()
-	}
-}
-
-// expireJob expires an action that hasn't progressed.
-func (s *server) expireJob(hash string) {
-	time.Sleep(expiryTime + time.Duration(rand.Int63n(int64(expiryTime))))
-	if s.maybeExpireJob(hash, false) {
-		return
+func (s *server) periodicallyDeleteJobs() {
+	for range s.deleteJobsTicker.C {
+		for digest, job := range s.jobs {
+			if shouldDeleteJob(job) {
+				s.mutex.Lock()
+				defer s.mutex.Unlock()
```

Comment:
You shouldn't defer this here. The defer won't be run until the function exits so you'll deadlock this pretty quickly.
Either pull the inner part out to its own function or maybe just manually unlock when done (since the code inside it is very simple)

I think this is what the linter is trying to tell you about although I'm not sure why it doesn't comment  here

Reply:
Ahh of course! This was blindly pulled out of the old delete job, have updated to unlock once we have deleted from the map. Thanks